Dockerfile

```diff
@@ -18,161 +19,277 @@
+# Ubuntun build toolchain
+RUN apt update
+RUN DEBIAN_FRONTEND=noninteractive apt install -y --no-install-recommends \
+        build-essential \
+        dirmngr \
+		dpkg-dev \
+        git \
+		libbluetooth-dev \
+		libbz2-dev \
+		libc6-dev \
+		libexpat1-dev \
+		libffi-dev \
+        libgmp-dev \
+		libgdbm-dev \
+		liblzma-dev \
+        libmpdec-dev \
+		libncursesw5-dev \
+		libreadline-dev \
+		libsqlite3-dev \
+		libssl-dev \
+		make \
+        netbase \
+        openjdk-11-jdk \
+		tk-dev \
+        tzdata \
+        unzip \
+		uuid-dev \
+		xz-utils \
+		zlib1g-dev \
+    && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /work
+
+# Copy the necessary bash resource to have paths
+# for different languages
+COPY docker/root-bashrc /root/.bashrc
+RUN mkdir -p /root/.bash_includes
+
+#------------------------------------------------------------------------
+# Python using pyenv
+ARG CONAN_VERSION=1.38.0
+ARG PYTHON_VERSION=3.8.11
+ARG SCANCODE_VERSION=21.8.4
+ARG VIRTUALENV_VERSION=20.2.2
 
+# Python pyenv
+ENV PYENV_ROOT=/opt/python
+RUN curl -L https://github.com/pyenv/pyenv-installer/raw/master/bin/pyenv-installer | bash
+ENV PATH=/opt/python/bin:$PATH
+RUN pyenv install "${PYTHON_VERSION}"
+RUN pyenv global "${PYTHON_VERSION}"
+ENV PATH=/opt/python/shims:$PATH
+RUN pip install -U pip
+RUN pip install conan=="${CONAN_VERSION}" \
+    Mercurial \
+    scancode-toolkit[full]=="${SCANCODE_VERSION}" \
+    virtualenv=="${VIRTUALENV_VERSION}" \
+    pipenv
+COPY docker/python.sh /root/.bash_includes
+
+#------------------------------------------------------------------------
+# Nodejs Using nvm
+ENV NVM_DIR=/opt/nodejs
+RUN git clone https://github.com/nvm-sh/nvm.git /opt/nodejs
+RUN . $NVM_DIR/nvm.sh && nvm install --lts
+COPY docker/nodejs.sh /root/.bash_includes
+
+#------------------------------------------------------------------------
+# Golang
+# Golang dep depends on some development packages to be installed, so need build
+# in the build stage
+ARG GO_DEP_VERSION=0.5.4
+ARG GO_VERSION=1.16.6
+ENV GOPATH=/opt/go
+RUN wget -qO- https://golang.org/dl/go${GO_VERSION}.linux-amd64.tar.gz | tar -C /opt -xz
+ENV PATH=/opt/go/bin:$PATH
+RUN go version
+RUN curl -ksS https://raw.githubusercontent.com/golang/dep/v$GO_DEP_VERSION/install.sh | bash
+RUN echo "add_local_path /opt/go/bin:$PATH" > /root/.bash_includes/go.sh
+
+#------------------------------------------------------------------------
+# Haskell
+ARG HASKELL_STACK_VERSION=2.1.3
+ENV DEST=/opt/haskell/bin/stack
+RUN curl -ksS https://raw.githubusercontent.com/commercialhaskell/stack/v$HASKELL_STACK_VERSION/etc/scripts/get-stack.sh | bash -s -- -d /usr/bin
+
+
+#------------------------------------------------------------------------
+# Ruby using rbenv
+ARG RUBY_VERSION=2.7.4
+ENV RBENV_ROOT=/opt/rbenv
+ENV PATH=${RBENV_ROOT}/bin:${RBENV_ROOT}/shims/:${RBENV_ROOT}/plugins/ruby-build/bin:$PATH
+RUN git clone https://github.com/rbenv/rbenv.git ${RBENV_ROOT}
```

Comment:
Instead of full clones, it might be worth to maybe do shallow clones?

Reply:
Done.Use shallow copies now